### PR TITLE
Added Username and Password for MQTT

### PR DESCRIPTION
**Pull request by nmaas87**
I added Username and Password for basic MQTT authentication for security reasons (so that we can at least use an basic protection on our server and the acl features of MQTT).

### CODE REVIEW

**Comment by nmaas87** [comment]
Sorry for the second pull request, I just cleaned up the PR a bit, the old one was very messy :(

File: CMakeLists.txt [hunk 1]
```diff
@@ -4,7 +4,7 @@ project(vallumd C)
 
 set (VERSION_MAJOR 0)
 set (VERSION_MINOR 1)
-set (VERSION_PATCH 1)
+set (VERSION_PATCH 2)
```
Comment (stintel) [inline CMakeLists.txt]: Please don't bump VERSION_PATCH. I'll decide that, and it always has to go in a commit on its own.

File: README.md [hunk 2]
```diff
@@ -64,8 +64,7 @@ Tested on CentOS 6 and 7.
 
 You can generate a DEB package with cpack:
 ```
-sudo apt-get -y install build-essential cmake libipset-dev libmosquitto-dev
-
+sudo apt-get -y install build-essential cmake libipset-dev libmosquitto-dev pkg-config
```
Comment (stintel) [inline README.md]: This change is unrelated, and should go in another commit. Ideally also add it for the CentOS/RedHat instructions.

File: src/main.c [hunk 3]
```diff
@@ -28,21 +28,25 @@
...
 {
     char *host = NULL;
+    char *username = NULL;
+    char *password = NULL;
     int opt = 0;
     unsigned int port = 1883;
     unsigned int t = 0;
 
-    while ((opt = getopt(argc, argv, "h:p:t:V")) != -1) {
+    while ((opt = getopt(argc, argv, "h:p:u:P:t:V")) != -1) {
```
Comment (stintel) [inline src/main.c]: Minor nit, but he options were in alphabetical order, please keep it that way.

File: src/main.c [hunk 4]
```diff
@@ -51,14 +55,20 @@ int main(int argc, char **argv)
     mqtt_topics = malloc(ntopics * sizeof(*mqtt_topics));
 
     optind = 0;
-    while ((opt = getopt(argc, argv, "h:p:t:V")) != -1) {
+    while ((opt = getopt(argc, argv, "h:p:u:P:t:V")) != -1) {
```
Comment (stintel) [inline src/main.c]: Minor nit, but he options were in alphabetical order, please keep it that way.

File: src/main.c [hunk 5]
```diff
@@ -51,14 +55,20 @@ int main(int argc, char **argv)
...
             case 'p':
                 port = atoi(optarg);
                 break;
+            case 'u':
+                username = optarg;
+                break;
+            case 'P':
+                password = optarg;
+                break;
             case 't':
```
Comment (stintel) [inline src/main.c]: Same here. Maintain alphabetical order.

File: src/mosquitto.c [hunk 6]
```diff
@@ -79,6 +81,7 @@ int init_mqtt()
     mosquitto_connect_callback_set(m, cb_con);
     mosquitto_message_callback_set(m, cb_msg);
 
+    mosquitto_username_pw_set(m, mqtt_username, mqtt_password);
```
Comment (stintel) [inline src/mosquitto.c]: Please group it with the other _set functions.

**Comment by stintel** [comment]
Also FYI, as this is a new feature, this is not going to be in 0.1.*.

**Comment by nmaas87** [comment]
Thank you for your review, I made the requested changes

**Comment by stintel** [comment]
Thanks. Some more comments, though:
* Please squash the first 3 commits.
* README.md vs README.MD
* drop the caps after the subject prefix
* imperative form (add, not added)
* last commit message makes no sense, it should just say what it does, the fact I requested changes doesn't belong there
* if it's possible to explain what the patch does just in the subject, please do so

Forgive me for making a problem out of all this, but I'm forcing myself to follow the same rules. I'll try to write contribution guidelines at some point.

So to avoid another "changes requested", I would write the subjects like this:

README.md: add missing pkg-config dependency
mosquitto: add username and password option

Says exactly what it does, no further explanation needed.
Thanks.

File: README.md [hunk 7]
```diff
@@ -49,7 +49,7 @@ included in many distributions yet.
 You can generate an RPM package with cpack:
 ```
 sudo wget -P /etc/yum.repos.d/ http://download.opensuse.org/repositories/home:/oojah:/mqtt/CentOS_CentOS-$(awk -v RS=[0-9]+ '{print RT+0;exit}' /etc/redhat-release)/home:oojah:mqtt.repo
-sudo yum -y install cmake ipset-devel libmosquitto-devel '@Development Tools'
+sudo yum -y install pkg-config cmake ipset-devel libmosquitto-devel '@Development Tools'
```
Comment (stintel) [inline README.md]: Alphabetical order please.

**Comment by nmaas87** [comment]
I tried to fix it with interactive Rebase, but somehow this brought more problems than thought. I just decided to cut all this stuff out and force push it. Should be ok now.

**Comment by nmaas87** [comment]
Still had a mix-up in the print_usage() section so that argument names were not sorted in right order, fixed now :)

**Comment by stintel** [comment]
Wasn't sure if I wanted it like that. It also makes sense to group username and password together. Oh well, let's just stick with alphabetical. Thanks!

**Comment by stintel** [comment]
@nmaas87 OK so I did something I'm not supposed to do, I rewrote history of master because I noticed two problems:
* author field does not match signed-off-by line
* there should *always* be an empty line after the subject

Can you resubmit? I think this will have to be a new PR, as this one is closed.
